### ISSUE TITLE
Suggestion: n run should download first

### ISSUE THREAD

**Issue by jaymefSO** [tj/n#647]
I think it would be handy if `n run` could download the version first before executing, instead of relying on the cached binary being present. This would be handy if you are trying to use `n run engine` to use the matching version from package.json (but may not always exist in cache). 

Why can't `n run` act similarly to how `n use` works to download the version first before setting it as default binary?

**Comment by shadowspawn** [comment]
I can see it would be a convenience, and work fine for a setup where install does not need any options or sudo, which is probably the most common. But in general, installing and using node have different concerns and options. For example, I may need to use sudo to install node. Or of current interest with the Apple M1 architecture, specify the architecture for the download. Or whether to use `xz` for the archive.

`n use` is just an alias for `n run`, perhaps you meant `n install` in your last comment?

Using `run` with `package.json` is an interesting case.

Without changes to `n`, you could ensure the version has been downloaded yourself (in a script or alias):

```
n --download engine
n run engine foo
```

**Comment by jaymefSO** [comment]
Ya, your suggestion is basically the only solution I've thought of so far. `n -d engine && n run engine foo`

The use case is a bit complex, but essentially it's a staging server which hosts many different projects. Traditionally they've always used the same version of Node, but now we require different versions for different projects. 

We run some node backend services using systemd units, using n with systemd poses some challenges, mostly surrounding minor version updates. I figured if we could include the engine version in package.json we could start the service using n run, the problem is that the version already has to be downloaded/cached for that to work. Downloading it first like suggested above _should_ do the job, but it would be nice if N had an option to force the download if it doesn't exist, or have n run work with the -d flag.

**Comment by shadowspawn** [comment]
(Thanks for the info on the context, interesting.)

**Comment by jaymefSO** [comment]
I found an easier way that I could potentially handle this use case with systemd by using ExecStartPre in the systemd unit

ex:

`ExecStartPre=/usr/bin/n -d engine
ExecStart=/usr/bin/n exec engine yarn start`

**Comment by shadowspawn** [comment]
Nice! Matches the two steps very well.

**Comment by shadowspawn** [comment]
The underlying problem was solved, and no further activity in a month. Closing this as resolved.

Feel free to open a new issue if it comes up again, with new information and renewed interest.

**Comment by mirao** [comment]
I'm using NodeJS 10 (by command `node`) by default and sometimes I want to use the latest NodeJS 12: `n run 12` or `n exec 12 npx codeceptjs`
Time from time (maybe when a new 12.x minor release is out) I get this:
`Error: '12' (12.22.6) not in downloads cache`
and I have to download the version manually:  `sudo n -d 12`

Would be great if `n run` and `n exec` did it automatically (or using -d switch)

**Comment by jaymefSO** [comment]
> I'm using NodeJS 10 (by command `node`) by default and sometimes I want to use the latest NodeJS 12: `n run 12` or `n exec 12 npx codeceptjs`
> Time from time (maybe when a new 12.x minor release is out) I get this:
> `Error: '12' (12.22.6) not in downloads cache`
> and I have to download the version manually: `sudo n -d 12`
> 
> Would be great if `n run` and `n exec` did it automatically (or using -d switch)

I agree, it would definitely be handy. We use `n exec` everywhere but we always have to run `n -d engine` before it because if the node version isn't already cached` n exec` will fail.

**Comment by shadowspawn** [comment]
Interesting to hear that `n exec` is getting used.

@mirao 
Your example shows why it can't just work transparently. With your current setup, are you going to put `sudo` in front of all your `n exec` and `n run` commands? (Of course not!) 

```
n exec 12 npx codeceptjs
# vs
sudo n -d 12
```

> Would be great if n run and n exec did it automatically (or using -d switch)

I am more comfortable making it an opt-in. I was going to argue that the `-d` belonged to `install` and not to `exec`, but actually that is an internal detail and not suggested by the help! Using `-d` does match the semantics.

@jaymefSO

> We use n exec a lot in CI pipelines and it's cumbersome having to make sure the node version is in download cache before running any n exec commands.

Hmm, I was leaving that up to user but hadn't thought about how to do it. One approach

```
n which 12 || n -d 12
```

**Comment by shadowspawn** [comment]
Expanded scripting for special targets to conditionally download node. (Not claiming it is is as convenient as built-in support would be, and not error checking for failed `lsr`!)
```
ENGINE=$(n lsr engine)
n which $ENGINE || n -d $ENGINE
n exec $ENGINE node --version
```

> All I'm saying is that it would be nice

Yes, and thanks for the comments. I have reopened issue for further consideration. 🤔

**Comment by jaymefSO** [comment]
hmm, when I use `n lsr engine` in a node project with engine specified in package.json I get this error:

`Error: invalid version 'engine'`

This is what is specified in package.json:
```

  "engines": {
    "node": "^14.15.1"
  },
```

**Comment by jaymefSO** [comment]
Scratch that, I upgraded from 7.2.2 to 7.3.1 and it worked with the newer release.

**Comment by shadowspawn** [comment]
Example output from experiment:

```
% node --version
v10.24.1
% n run lts --version

  Error: 'lts' (14.17.6) not in downloads cache

% n run -d lts --version
  installing : node-v14.17.6
       mkdir : /usr/local/n/versions/node/14.17.6
       fetch : https://nodejs.org/dist/v14.17.6/node-v14.17.6-darwin-x64.tar.xz
v14.17.6
% node --version
v10.24.1
```

**Comment by jaymefSO** [comment]
> Example output from experiment:
> 
> ```
> % node --version
> v10.24.1
> % n run lts --version
> 
>   Error: 'lts' (14.17.6) not in downloads cache
> 
> % n run -d lts --version
>   installing : node-v14.17.6
>        mkdir : /usr/local/n/versions/node/14.17.6
>        fetch : https://nodejs.org/dist/v14.17.6/node-v14.17.6-darwin-x64.tar.xz
> v14.17.6
> % node --version
> v10.24.1
> ```

This looks awesome, thanks so much for working on this!

**Comment by shadowspawn** [comment]
Released in `n` v7.4.0

**Comment by jaymefSO** [comment]
Thank you so much, we really appreciate this. I know my dev team will certainly be happy to hear that this enhancement has been implemented. Cheers!

**Comment by mirao** [comment]
Just tried it and it seems that `n -d exec 12 npx codeceptjs --version` downloads not only latest NodeJS 12, but also latest CodeceptJS 3.1.2:
```bash 
  installing : node-v12.22.6
       mkdir : /home/mirao/.n/n/versions/node/12.22.6
       fetch : https://nodejs.org/dist/v12.22.6/node-v12.22.6-linux-x64.tar.xz
npx: installed 250 in 23.836s
3.1.2
```

![Screenshot from 2021-09-10 22-36-01](https://user-images.githubusercontent.com/12584138/132914938-747c03e3-ae46-4c10-9a9f-3bdcb539e0af.png)

If I run it again, I get correct (my project's version) of CodeceptJS:
```bash
$ n -d exec 12 npx codeceptjs --version
3.0.7
```

Any idea what's wrong?

**Comment by shadowspawn** [comment]
> Any idea what's wrong?

No... no... Yes! 
Drat, working directory is wrong after the download. Bug fix coming.

**Comment by shadowspawn** [comment]
Released `n` v7.4.1 with fix for current directory being affected by download

**Comment by mirao** [comment]
@shadowspawn Thank you very much, the patched version works well
